Implement X500DistinguishedNameBuilder

vcsjones opened the pull request:
Closes #44738

vcsjones committed:
Initial implementation

vcsjones committed:
Tests

vcsjones committed:
Finish up tests

vcsjones committed:
Make Windows like me

vcsjones commented:
Draft until docs are done.

vcsjones committed:
Docs and better input validation

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs:
```diff
@@ -8,16 +8,71 @@
+        /// <paramref name="oidValue" /> or <paramref name="value" /> is <see langword="null" />.
+        /// </exception>
+        /// <exception cref="ArgumentException">
+        /// <p>
```
All of these `<p>`s (HTML) should be `<para>`s (xmldoc)

vcsjones committed:
p to para

bartonjs commented on src/libraries/System.Security.Cryptography/src/Resources/Strings.resx:
```diff
@@ -78,6 +78,15 @@
+  </data>
+  <data name="Argument_Asn1_InvalidDer" xml:space="preserve">
+    <value>The contents are not valid ASN.1 under Distinguished Encoding Rules.</value>
+  </data>
```
Please help quiet my neuroses and move these collectively down one (Arg_RankMultiDimNotSupported should be above them, Argument_DestinationTooShort below)

bartonjs commented on src/libraries/System.Security.Cryptography/src/Resources/Strings.resx:
```diff
@@ -93,6 +102,12 @@
+  <data name="Argument_X500_InvalidCountryOrRegion" xml:space="preserve">
+    <value>The country or region code must be exactly two characters.</value>
+  </data>
+  <data name="Argument_X500_EmailTooLong" xml:space="preserve">
```
Argument_X500 should be much lower down :)

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs:
```diff
@@ -0,0 +1,479 @@
+    /// <summary>
+    /// This class facilitates building a distinguished name for an X.509 certificate.
+    /// </summary>
+    public sealed class X500DistinguishedNameBuilder
```
If you can come up with a sane way to describe the order of operations here, that'd probably make for a good remarks section.

Assuming that we went with the user-sanity implementation over the code-sanity implementation, that'd be finding some way to explain something like

> For an X500DistinguishedName value whose Name property prints as "CN=widgets.example.org, OU=Widget Processing Center, O=Example Org" you would call AddCommonName("widgets.example.org"), then AddOrganizationalUnit("Widget Processing Center"), then AddOrganization("Example Org"), and finally Encode().

Or maybe it's only confusing when you're over-burdened with knowledge.  That this reverses, and X500DN reverses, so everything just works like anyone who doesn't know all this actually works thinks it does.

vcsjones replied:
I came up with a way, though I'll let you be the judge of it's sanity.

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs:
```diff
@@ -0,0 +1,479 @@
+            ReadOnlySpan<byte> value,
+            [CallerArgumentExpression("value")] string? valueParamName = null)
+        {
+            if (!AsnDecoder.TryReadEncodedValue(value, AsnEncodingRules.DER, out _, out _, out _, out _))
```
If skipping the length check is intentional (so that "invalid DER" and "one value at a time" are different messages) then that's probably worth a comment.

Otherwise, it feels like we should check that bytesConsumed == value.Length or throw the message controlled at this layer.

vcsjones replied:
I reversed course here after adding a few tests.

bartonjs commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,379 @@
+        public static void AddEmailAddress_ExceedsMaximumLength_Fails()
+        {
+            X500DistinguishedNameBuilder builder = new();
+            string bigEmail = new string('a', 244) + "@example.com";
```
We could make this test simpler, and avoid math, by just using `new string('a', 256)`, since the implementation doesn't look for `@`.

vcsjones replied:
It seemed more sensible to me to test the "real" thing, similar to your sentiment further below:

> I certainly don't want to make this a negative test, but, IIRC, it's supposed to be the full name of the state, not the abbreviation. So "California"

But we can add a test that gibberish is accepted.

bartonjs commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,379 @@
+                "30223120301E06092A864886F70D01090116116B6576696E406578616D706C652E636F6D",
+                builder => builder.AddEmailAddress("kevin@example.com"));
+        }
+
```
Another we-don't-care-about-rfc-822 comment: Probably worth a test that "flibberty gibbet" (or whatever suits you) is a valid email address, as far as the builder is concerned.

Has a space, missing an `@`.  But: totes valid.  Yep, I said "totes".

bartonjs commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,379 @@
+        {
+            AssertBuilder("300D310B3009060355040613025553", builder => builder.AddCountryOrRegion("US"));
+        }
+
```
Whatever we decide for non-CAPS-CAPS, we should have tests.

I think I'm leaning toward "normalize and filter".

* uS => US
* u1 => throw

if we go GIGO, then just the tests that show that.

bartonjs commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,379 @@
+            X500DistinguishedNameBuilder builder = new();
+            ArgumentException ex = AssertExtensions.Throws<ArgumentException>(
+                "twoLetterCode",
+                () => builder.AddCountryOrRegion("``"));
```
The comments called out a surrogate pair as an interesting case.  May as well throw one in for the tests.

bartonjs commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,379 @@
+        [Fact]
+        public static void AddStateOrProvinceName_Success_EncodeToUTF8String()
+        {
+            AssertBuilder("300D310B300906035504080C024341", builder => builder.AddStateOrProvinceName("CA"));
```
I certainly don't want to make this a negative test, but, IIRC, it's supposed to be the full name of the state, not the abbreviation.  So "California"

> 6.3.5 State or Province Name
The State or Province Name attribute type specifies a state or province. When used as a component of a directory name, it identifies a geographical subdivision in which the named object is physically located or with which it is associated in some other important way.
An attribute value for State or Province Name is a string, e.g., **S = "Ohio"**.
stateOrProvinceName ATTRIBUTE ::= {
SUBTYPE OF name
WITH SYNTAX UnboundedDirectoryString
LDAP-SYNTAX directoryString.&id
LDAP-NAME {"st"}
ID id-at-stateOrProvinceName }

(S = Ohio, emphasis mine)

bartonjs commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,379 @@
+                () => builder.Add((string)null, "banana"));
+            AssertExtensions.Throws<ArgumentException>(
+                "oidValue",
+                () => builder.Add("", "banana"));
```
I used to love "banana" just because of the "anana".  Now I've seen Horrible Histories and love it even more.

https://youtu.be/zPtYmq5qFVA?t=5.  You're welcome :)

bartonjs commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,379 @@
+        {
+            AssertBuilder(expectedHex, builder => builder.Add(new Oid(oid, null), value, tag));
+        }
+
```
A test that adds 2 or more things to show the ordering would be good.

And probably a second one that shows that duplicates are кosher.

vcsjones replied:
I see what you did there.

bartonjs commented:
Conversation from my head:

```
> Ask Kevin where the tests are for multi-valued.
< But the code didn't handle that.
> Yeah, that's why it's a good question.
< But the code didn't handle that.
> Are you listening to me?
< Yeah. The code doesn't handle it, I don't think it's possible. Maybe you want to rephrase the question.
> Fine, as Kevin where the escape valve for multi-valued RDNs is.
< Sure. I think it got cut.
> What?!
< Yeah. Something about anyone who would know what a multi-valued RDN is can probably figure out AsnWriter on their own.
> Oh, yeah. That sounds familiar.
< Happy now?
> Yeah.  Note to self, don't ask Kevin about multi-valued RDNs.
```

bartonjs commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs:
```diff
@@ -0,0 +1,479 @@
+
+            using (writer.PushSequence())
+            {
+                foreach (byte[] component in _encodedComponents)
```
I feel like this loop is backwards from what we wanted it to be.  Because strings and X500DN reversing and yadda yadda yadda.

bartonjs commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,379 @@
+        {
+            AssertBuilder(expectedHex, builder => builder.Add(new Oid(oid, null), value, tag));
+        }
+
```
A lifetime test is probably also useful.

* Add some stuff, call Build.
* Add more stuff.  Note how it doesn't throw.  Call Build.  See that it was cumulative.

vcsjones committed:
Order strings resources, for sanity purposes

vcsjones committed:
Use a better reference

vcsjones commented:
> Conversation from my head:

>> Ask Kevin where the tests are for multi-valued.

What a coincidence. I had the same conversation with the same guy earlier this morning.

vcsjones committed:
Stricter checking of country or region

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs:
```diff
@@ -0,0 +1,479 @@
+            {
+                throw new ArgumentException(SR.Argument_X500_InvalidCountryOrRegion, nameof(twoLetterCode));
+            }
+
```
> but it feels like we can be guiding and flexible by rejecting digits and punctuation.

I think that is a good balance of validating and not being too strict about inputs.

vcsjones committed:
Code review feedback.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs:
```diff
@@ -0,0 +1,479 @@
+            UniversalTagNumber stringEncodingType,
+            [CallerArgumentExpression("value")] string? paramName = null)
+        {
+            AsnWriter writer = new AsnWriter(AsnEncodingRules.DER);
```
I thought about this... and got a little unsure from exception handling.

Consider if someone did:

```C#
X500DistinguishedNameBuilder builder = new();

try
{
    builder.AddEmailAddress(<something that is not IA5>);
}
catch {}

// What is the state of the shared writer here?
```

If the writer is shared, we might end up with things left-over in the writer like the pushed SET OF+SEQUENCE OF.

_Kevin, is that something that really matters?_

I... think... so. The shared writer should not be in an invalid state after exceptions.

_Okay Kevin, but we can call `Clear` *before* every use to make sure we have a clean slate instead of trying to clean up after ourselves_

We could... and we need to make a test to make sure the builder is still in a valid state after an exception... and the first `Reset` would effectively a no-op if we do them up-front... but wait doesn't `Reset` internally zero buffers since AsnWriter doesn't know if it's dealing with sensitive data?

So I just ended up creating a new writer every time. If we think it's wasteful, I wouldn't mind doing the work.

bartonjs replied:
AsnWriter.Clear() does zero out the contents of the buffer... but only as much as it has written.  Since it doesn't use an array pool, it's always fetching a zeroed 1k array for a new one; so holding one instance and calling Clear is probably less overall work.

But, yeah, Clear-first (like ERR_clear_error() first) is an interesting, and useful, observation/pattern.

vcsjones commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,379 @@
+                "oidValue",
+                () => builder.Add("", "banana"));
+        }
+
```
Well... an empty subject [is legal](https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.6) for leaf certs, so I don't feel that we need to constrain it to something that is populated. My vote would be permitting it (with a test, of course).

bartonjs replied:
Seems like enough justification to me.

vcsjones committed:
Use a shared writer

vcsjones committed:
Test empty builder

vcsjones committed:
More code review feedback

vcsjones commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,379 @@
+            AssertExtensions.Throws<ArgumentException>(
+                "encodedValue",
+                () => builder.AddEncoded(new Oid(TestOid), stackalloc byte[] { 0xFF, 0xFF, 0xFF }));
+        }
```
> 01 01 01 (throw, not DER-canonical)

`AsnDecoder` won't throw for this case. This goes back to my question in https://github.com/dotnet/runtime/issues/45169, which is basically that `AsnDecoder` doesn't really validate the contents of UNIVERSAL types.

I generally think this is okay, since `AddEncoded` is an escape hatch for "I know what I am doing". If we feel differently about that, I would be more interested in how we can improve the System.Formats.Asn1 functionality to do more DER validation and this can leverage that if/when that capability is added.

vcsjones replied:
Though, I would probably point out that if we don't restrict it now, it would be hard to further restrict it later. However, I still think the escape-hatch-esque utility for this still applies.

vcsjones replied:
Okay - wait. The ASN.1 APIs are still a little new to me. There is validation... I just need to actually move though it.

vcsjones replied:
I looked further in to this and... I generally feel that this is non-trivial and that something is missing from AsnReader, AsnDecoder, or something to further validate ASN.1 is actually DER by reading the whole blob. I went down the path of trying to validate the whole document here:

https://gist.github.com/vcsjones/4892cfeca11bfe9fb135772afdf9f35e

and started wondering if I am missing something with System.Formats.Asn1?

bartonjs replied:
The Remarks section of TryReadEncodedValue does say that it doesn't validate a lot (mainly that DER can't use indefinite encodings, etc).  https://docs.microsoft.com/en-us/dotnet/api/system.formats.asn1.asndecoder.tryreadencodedvalue?view=net-6.0#remarks

Perhaps we want an AsnDecoder::TryValidateEncodedValue, which would be the recursive thing you started writing there (and I think we have somewhere in CertificateRequest)

vcsjones commented:
@bartonjs thank you for the thorough feedback. I think I have addressed all of your feedback, with the exception of "Is what `AddEncoded` taking really DER?" which, to my understanding, is non-trivial to do correctly, unless I am missing something. If I am not missing something, we can:

1. Allow `AddEncoded` to be relaxed, as its current state, with the validation being limited to length encoding and no trailing data. We take the stance that this API is meant to take pre-encoded, pre-validated data, and document it as such.
2. Block this PR on new functionality in `System.Formats.Asn1` that does the DER validation we want.
3. Add the DER validation in `X500DistinguishedNameBuilder`.
4. Remove `AddEncoded` from the new API set until we have some validation available to us, if ever.

My order of preference is 1, 4, 2, 3.

vcsjones committed:
Additional code review feedback

bartonjs commented:
Ignoring my `01 01 01` testcase, as that ended up worrying about the payload, I think the subconscious intent behind my question was really more about the acceptable tags.

https://github.com/dotnet/runtime/blob/39fb7f7826270f00b856e3e9a13165c07dec87cc/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DictionaryStringHelper.cs limits the outputs to Universal-class character string values.  Reading through https://datatracker.ietf.org/doc/html/rfc4517 shows that almost everything looks like text... until I realized that it is also providing textual output for non-textual attribute values... so I think that means our X500DN.Format is missing a bunch of RFC 4517 rules (and probably Windows is, too, or maybe we just need some more test cases to improve our compatibility if they have more support than I recall).

Forcing something to use a particular encoding, e.g. OU to use BMPEncoding, is solvable by Add.  I forget why we wanted AddEncoded... maybe it was exactly for "this is something other than a text string" (unless it was "I'm walking through things using AsnReader, but want to replace one element... here, this is already encoded, just take it... don't make me turn it into a string for you to turn it back into bytes).  I think I've talked myself in a circle as to whether we want to limit the type to character strings only or not.

vcsjones commented:
> forget why we wanted AddEncoded... maybe it was exactly for "this is something other than a text string"

I think there was more utility back when we thought we would have multi-RDN value support.

I don’t recall ever seeing non-text inputs, even for things where they could be better-than-text, they are usually strings. e.g. I’ve seen OIDs in subjects, but they are an OID as a PrintableString, not an OID as an OBJECT IDENTIFIER.

I suppose `AddEncoded` could be removed under the same pretense that multi-RDN was removed: you are doing something interesting / unique and you can probably just use `AsnWriter`.

bartonjs commented:
> I think there was more utility back when we thought we would have multi-RDN value support.

Yeah, but we have the wrong shape for that.  The one multi-valued test we have (https://github.com/dotnet/runtime/blob/e74c3e812fa5c96dc7ebaff655bff8820f3f9a62/src/libraries/System.Security.Cryptography.X509Certificates/tests/X500DistinguishedNameTests.cs#L23-L33) shows the encoding:

```
SEQUENCE (1 elem)
  SET (2 elem)
    SEQUENCE (2 elem)
      OBJECT IDENTIFIER 2.5.4.3 commonName (X.520 DN component)
      PrintableString James
    SEQUENCE (2 elem)
      OBJECT IDENTIFIER 2.5.4.10 organizationName (X.520 DN component)
      PrintableString Microsoft
```

If it was a multi-RDN escape hatch then we'd've had just `AddEncoded` which required the input to be a SET( OF).

> I suppose AddEncoded could be removed under the same pretense that multi-RDN was removed: you are doing something interesting / unique and you can probably just use AsnWriter.

Yeah.  Let's just cut it.  If we remember why we wanted it it's easy enough to add back (and already API-Approved)

bartonjs commented on src/libraries/System.Security.Cryptography/tests/X509Certificates/X500DistinguishedNameBuilderTests.cs:
```diff
@@ -0,0 +1,452 @@
+                builder.AddCountryOrRegion("US");
+                builder.AddOrganizationName("GitHub");
+                builder.AddOrganizationalUnitName("PSE");
+                builder.AddDomainComponent("github");
+                builder.AddDomainComponent("com");
```
I was about to say "it looks like we're missing the test that shows the same attribute type ID can be used more than once", but KitchenSink adds DC twice, so it's covered.

vcsjones committed:
Remove AddEncoded

vcsjones commented:
> If it was a multi-RDN escape hatch

I didn’t intend to imply that I thought it was doing that, just that back when we were shaping out the API, it was an escape hatch for that. But then we threw an OID on there (so it can’t take the SET OF) so it didn’t make sense and we should have cut it instead.

vcsjones commented on src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X500DistinguishedNameBuilder.cs:
```diff
@@ -0,0 +1,409 @@
+        /// </exception>
+        public void Add(string oidValue, string value!!, UniversalTagNumber? stringEncodingType = null)
+        {
+            ArgumentException.ThrowIfNullOrEmpty(oidValue);
```
It was something I considered and ultimately decided the order in which parameters are checked for not-null wasn't something that strongly warranted changing anything. I did not write tests for this behavior though because I don't think it makes sense to write a test that locks us in to a specific throwing order for `ArgumentNullException`. The tests as written in such a way that each parameter is tested independently.
